File: java/src/mealtrainer/core/DeceleratedCurve.java
package mealtrainer.core;

import java.lang.Math;
import java.lang.Double;
import mealtrainer.core.util.Util;
import mealtrainer.core.MealConfig;

public class DeceleratedCurve {

     private final MealConfig config;
     private final double rateOfDeceleration;
     private final double initialEatingRate;

     public DeceleratedCurve(MealConfig config) {
          this.config = config;
          this.initialEatingRate = findInitialRateForMinimalTargetDifference();
          this.rateOfDeceleration = getRateOfDecline(initialEatingRate, config.getMealDuration());
     }

     public int getCurrentSecondsTillNextBite(int currentTimeSeconds) {
          final double decimalMinutes = Double.valueOf(currentTimeSeconds) / 60.0;
          final double estimateFoodConsumed = estimateWeightConsumedAtTime(decimalMinutes);
          final double foodConsumedIn1Minute = estimateWeightConsumedAtTime(decimalMinutes + 1);
          final double consumptionDelta = foodConsumedIn1Minute - estimateFoodConsumed;
          final double bitesNeeded = Math.max(consumptionDelta / config.weightPerBite(), 1.0);
          final double timePerBite = 60.0 / bitesNeeded;
          return (int) timePerBite;
     }

     private double estimateWeightConsumedAtTime(double x) {
          return (rateOfDeceleration * (x * x)) + (initialEatingRate * x);
     }

     private double getRateOfDecline(double initialEatingRate, int durationMinutes) {
          return -1 * (initialEatingRate / (2 * durationMinutes));
     }

     private double findInitialRateForMinimalTargetDifference() {
          final double[] B = Util.linspace(5, 160, 100);
          final double[] diffs = getDiffs(B);
          final double min_diff = Util.minValue(diffs);
          if (min_diff > 1) {
               System.err.println(
                         "[WARNING]: Can't estimate the target weight to within 1g.\n\t   Minimum absolute difference to target weight '"
                                   + min_diff + "g' for duration '" + config.getMealDuration() + " min' is "
                                   + config.getTargetWeight() + "g.");
          }
          return B[Util.indexOf(diffs, min_diff)];
     }

     private double[] getDiffs(double[] bValues) {
          double[] out = new double[bValues.length];
          final int target = config.getTargetWeight();
          final int duration = config.getMealDuration();
          for (int i = 0; i < bValues.length; i++) {
               out[i] = target
                         - (getRateOfDecline(bValues[i], duration) * (duration * duration) + bValues[i] * duration);
          }
          return out;
     }

}

File: python/decelerated_curve.py
import utils


class DeceleratedCurve():

    def __init__(self, config):
        self.config = config
        self.b, _ = self.__find_b_for_minimal_target_difference(
            config.target_weight, config.duration)
        self.a = self.__get_deceleration(self.b, config.duration)
        self.default_time_space = utils.construct_value_space(
            0, config.duration)

    def get_quadratic_curve_shape(self, time_space=None):
        time = self.default_time_space if time_space != None else time_space
        return list(map(lambda x: self.__estimate_weight_consumed_at_time(x), time))

    def get_time_till_next_bite(self, current_time_secs):
        x = utils.to_decimal_minutes(current_time_secs)
        estimate_food_consumed = self.__estimate_weight_consumed_at_time(x)
        food_consumed_1_min_ahead = self.__estimate_weight_consumed_at_time(
            x + 1)
        consumption_delta = food_consumed_1_min_ahead - estimate_food_consumed
        bites_needed = max(consumption_delta / self.config.weight_per_bite, 1)
        time_per_bite = 60.0 / bites_needed
        return time_per_bite

    def __find_b_for_minimal_target_difference(self, target, duration):
        B = utils.construct_value_space(5, 160)
        diffs = list(map(lambda b: abs(
            target - (self.__get_deceleration(b, duration) * duration ** 2 + b * duration)), B))
        min_diff = min(diffs)
        if min_diff > 1:
            print("**********************************************************")
            print("[WARNING]: Can't estimate the target weight to within 1g.\n\t   Minimum absolute difference to target weight '{0}g' for duration '{1} min' is {2}g.".format(
                target, duration, min_diff))
            print("**********************************************************")
        return B[diffs.index(min_diff)], min_diff

    def __estimate_weight_consumed_at_time(self, x):
        return (self.a * x ** 2) + (self.b * x)

    def __get_deceleration(self, eating_rate, duration):
        return -(eating_rate/(2*duration))
